// cloudflared/lib/lib.js
// node_modules/.pnpm/tsup@6.1.3_typescript@4.7.4/node_modules/tsup/assets/esm_shims.js
import { fileURLToPath } from "url";
import path from "path";
var getFilename = () => fileURLToPath(import.meta.url);
var getDirname = () => path.dirname(getFilename());
var __dirname = /* @__PURE__ */ getDirname();

// src/constants.ts
import path2 from "path";
var bin = path2.join(__dirname, "..", "bin", process.platform === "win32" ? "cloudflared.exe" : "cloudflared");

// src/install.ts
import fs from "fs";
import path3 from "path";
import https from "https";
import { execSync } from "child_process";
var LINUX_URL = {
  arm64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-arm64",
  arm: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-arm",
  x64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-amd64",
  ia32: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-386"
};
var MACOS_URL = {
  arm64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-darwin-amd64.tgz",
  x64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-darwin-amd64.tgz"
};
var WINDOWS_URL = {
  x64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-windows-amd64.exe",
  ia32: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-windows-386.exe"
};
async function install(to) {
  if (process.platform === "linux") {
    return install_linux(to);
  } else if (process.platform === "darwin") {
    return install_macos(to);
  } else if (process.platform === "win32") {
    return install_windows(to);
  } else {
    console.error("Unsupported platform: " + process.platform);
    process.exit(1);
  }
}
async function install_linux(to) {
  const url = LINUX_URL[process.arch];
  if (url === void 0) {
    console.error("Unsupported architecture: " + process.arch);
    process.exit(1);
  }
  await download(url, to);
  await new Promise((r) => setTimeout(r, 100));
  fs.chmodSync(to, "755");
  return to;
}
async function install_macos(to) {
  const url = MACOS_URL[process.arch];
  if (url === void 0) {
    console.error("Unsupported architecture: " + process.arch);
    process.exit(1);
  }
  await download(url, `${to}.tgz`);
  process.env.VERBOSE && console.log(`Extracting to ${to}`);
  execSync(`tar -xzf ${path3.basename(`${to}.tgz`)}`, { cwd: path3.dirname(to) });
  fs.unlinkSync(`${to}.tgz`);
  fs.renameSync(`${path3.dirname(to)}/cloudflared`, to);
  return to;
}
async function install_windows(to) {
  const url = WINDOWS_URL[process.arch];
  if (url === void 0) {
    console.error("Unsupported architecture: " + process.arch);
    process.exit(1);
  }
  await download(url, to);
  return to;
}
function download(url, to, redirect = 0) {
  if (redirect === 0) {
    process.env.VERBOSE && console.log(`Downloading ${url} to ${to}`);
  } else {
    process.env.VERBOSE && console.log(`Redirecting to ${url}`);
  }
  return new Promise((resolve, reject) => {
    if (!fs.existsSync(path3.dirname(to))) {
      fs.mkdirSync(path3.dirname(to), { recursive: true });
    }
    let done = true;
    const file = fs.createWriteStream(to);
    const request = https.get(url, (res) => {
      if (res.statusCode === 302 && res.headers.location !== void 0) {
        done = false;
        file.close();
        resolve(download(res.headers.location, to, redirect + 1));
        return;
      }
      res.pipe(file);
    });
    file.on("finish", () => {
      if (done) {
        resolve(to);
      }
    });
    request.on("error", (err) => {
      fs.unlink(to, () => reject(err));
    });
    file.on("error", (err) => {
      fs.unlink(to, () => reject(err));
    });
    request.end();
  });
}

// src/utils.ts
import { spawn } from "child_process";
function tunnel(options = {}) {
  const args = ["tunnel"];
  for (const [key, value] of Object.entries(options)) {
    if (typeof value === "string") {
      args.push(`--${key}`, value);
    } else if (typeof value === "number") {
      args.push(`--${key}`, value.toString());
    } else if (value === null) {
      args.push(`--${key}`);
    }
  }
  const child = spawn(bin, args, { stdio: ["ignore", "pipe", "pipe"] });
  if (process.env.VERBOSE) {
    child.stdout.pipe(process.stdout);
    child.stderr.pipe(process.stderr);
  }
  const url_regex = /\|\s+(https?:\/\/[^\s]+)/;
  let url_resolver = () => void 0;
  let url_rejector = () => void 0;
  const url = new Promise((...pair) => [url_resolver, url_rejector] = pair);
  const connection_regex = /Connection ([a-z0-9-]+) registered connIndex=(\d) ip=([0-9.]+) location=([A-Z]+)/;
  const connection_resolvers = [];
  const connection_rejectors = [];
  const connections = [];
  for (let i = 0; i < 4; i++) {
    connections.push(new Promise((...pair) => [connection_resolvers[i], connection_rejectors[i]] = pair));
  }
  const parser = (data) => {
    const str = data.toString();
    const url_match = str.match(url_regex);
    url_match && url_resolver(url_match[1]);
    const connection_match = str.match(connection_regex);
    if (connection_match) {
      const [, id, idx, ip, location] = connection_match;
      connection_resolvers[+idx]({ id, ip, location });
    }
  };
  child.stdout.on("data", parser).on("error", url_rejector);
  child.stderr.on("data", parser).on("error", url_rejector);
  const stop = () => child.kill("SIGINT");
  return { url, connections, child, stop };
}
export {
  bin,
  install,
  tunnel
};
