// cloudflared/lib/lib.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/lib.ts
var lib_exports = {};
__export(lib_exports, {
  bin: () => bin,
  install: () => install,
  tunnel: () => tunnel
});
module.exports = __toCommonJS(lib_exports);

// src/constants.ts
var import_node_path = __toESM(require("path"), 1);
var bin = import_node_path.default.join(__dirname, "..", "bin", process.platform === "win32" ? "cloudflared.exe" : "cloudflared");

// src/install.ts
var import_node_fs = __toESM(require("fs"), 1);
var import_node_path2 = __toESM(require("path"), 1);
var import_node_https = __toESM(require("https"), 1);
var import_node_child_process = require("child_process");
var LINUX_URL = {
  arm64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-arm64",
  arm: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-arm",
  x64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-amd64",
  ia32: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-386"
};
var MACOS_URL = {
  arm64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-darwin-amd64.tgz",
  x64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-darwin-amd64.tgz"
};
var WINDOWS_URL = {
  x64: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-windows-amd64.exe",
  ia32: "https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-windows-386.exe"
};
async function install(to) {
  if (process.platform === "linux") {
    return install_linux(to);
  } else if (process.platform === "darwin") {
    return install_macos(to);
  } else if (process.platform === "win32") {
    return install_windows(to);
  } else {
    console.error("Unsupported platform: " + process.platform);
    process.exit(1);
  }
}
async function install_linux(to) {
  const url = LINUX_URL[process.arch];
  if (url === void 0) {
    console.error("Unsupported architecture: " + process.arch);
    process.exit(1);
  }
  await download(url, to);
  await new Promise((r) => setTimeout(r, 100));
  import_node_fs.default.chmodSync(to, "755");
  return to;
}
async function install_macos(to) {
  const url = MACOS_URL[process.arch];
  if (url === void 0) {
    console.error("Unsupported architecture: " + process.arch);
    process.exit(1);
  }
  await download(url, `${to}.tgz`);
  process.env.VERBOSE && console.log(`Extracting to ${to}`);
  (0, import_node_child_process.execSync)(`tar -xzf ${import_node_path2.default.basename(`${to}.tgz`)}`, { cwd: import_node_path2.default.dirname(to) });
  import_node_fs.default.unlinkSync(`${to}.tgz`);
  import_node_fs.default.renameSync(`${import_node_path2.default.dirname(to)}/cloudflared`, to);
  return to;
}
async function install_windows(to) {
  const url = WINDOWS_URL[process.arch];
  if (url === void 0) {
    console.error("Unsupported architecture: " + process.arch);
    process.exit(1);
  }
  await download(url, to);
  return to;
}
function download(url, to, redirect = 0) {
  if (redirect === 0) {
    process.env.VERBOSE && console.log(`Downloading ${url} to ${to}`);
  } else {
    process.env.VERBOSE && console.log(`Redirecting to ${url}`);
  }
  return new Promise((resolve, reject) => {
    if (!import_node_fs.default.existsSync(import_node_path2.default.dirname(to))) {
      import_node_fs.default.mkdirSync(import_node_path2.default.dirname(to), { recursive: true });
    }
    let done = true;
    const file = import_node_fs.default.createWriteStream(to);
    const request = import_node_https.default.get(url, (res) => {
      if (res.statusCode === 302 && res.headers.location !== void 0) {
        done = false;
        file.close();
        resolve(download(res.headers.location, to, redirect + 1));
        return;
      }
      res.pipe(file);
    });
    file.on("finish", () => {
      if (done) {
        resolve(to);
      }
    });
    request.on("error", (err) => {
      import_node_fs.default.unlink(to, () => reject(err));
    });
    file.on("error", (err) => {
      import_node_fs.default.unlink(to, () => reject(err));
    });
    request.end();
  });
}

// src/utils.ts
var import_node_child_process2 = require("child_process");
function tunnel(options = {}) {
  const args = ["tunnel"];
  for (const [key, value] of Object.entries(options)) {
    if (typeof value === "string") {
      args.push(`--${key}`, value);
    } else if (typeof value === "number") {
      args.push(`--${key}`, value.toString());
    } else if (value === null) {
      args.push(`--${key}`);
    }
  }
  const child = (0, import_node_child_process2.spawn)(bin, args, { stdio: ["ignore", "pipe", "pipe"] });
  if (process.env.VERBOSE) {
    child.stdout.pipe(process.stdout);
    child.stderr.pipe(process.stderr);
  }
  const url_regex = /\|\s+(https?:\/\/[^\s]+)/;
  let url_resolver = () => void 0;
  let url_rejector = () => void 0;
  const url = new Promise((...pair) => [url_resolver, url_rejector] = pair);
  const connection_regex = /Connection ([a-z0-9-]+) registered connIndex=(\d) ip=([0-9.]+) location=([A-Z]+)/;
  const connection_resolvers = [];
  const connection_rejectors = [];
  const connections = [];
  for (let i = 0; i < 4; i++) {
    connections.push(new Promise((...pair) => [connection_resolvers[i], connection_rejectors[i]] = pair));
  }
  const parser = (data) => {
    const str = data.toString();
    const url_match = str.match(url_regex);
    url_match && url_resolver(url_match[1]);
    const connection_match = str.match(connection_regex);
    if (connection_match) {
      const [, id, idx, ip, location] = connection_match;
      connection_resolvers[+idx]({ id, ip, location });
    }
  };
  child.stdout.on("data", parser).on("error", url_rejector);
  child.stderr.on("data", parser).on("error", url_rejector);
  const stop = () => child.kill("SIGINT");
  return { url, connections, child, stop };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  bin,
  install,
  tunnel
});
