// cloudflared-tunnel/dist/services/cloudflared-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installBinary = exports.stopTunnel = exports.startTunnel = exports.routeDnsToTunnel = exports.login = exports.createTunnel = exports.isRemoteTunnelExist = exports.updateConfig = exports.isAllReady = exports.isTunnelCreated = exports.isCredentialsFilesExist = exports.getConfig = exports.resetConfig = exports.isInstalled = exports.CONFIG_FILE = exports.CLOUDFLARED_USR_DIR = exports.TUNNELNAME = void 0;
const core_1 = require("@oclif/core");
const cloudflared_1 = require("cloudflared");
const node_child_process_1 = require("node:child_process");
const node_fs_1 = require("node:fs");
const node_os_1 = require("node:os");
const node_path_1 = require("node:path");
const node_process_1 = require("node:process");
const readline_sync_1 = require("readline-sync");
let stop;
exports.TUNNELNAME = 'cflared';
exports.CLOUDFLARED_USR_DIR = (0, node_path_1.resolve)((0, node_os_1.homedir)(), '.cloudflared');
exports.CONFIG_FILE = (0, node_path_1.resolve)(exports.CLOUDFLARED_USR_DIR, exports.TUNNELNAME);
function isInstalled() {
    node_process_1.stdout.write('Checking for cloudflared binary...\n');
    return (0, node_fs_1.existsSync)(cloudflared_1.bin);
}
exports.isInstalled = isInstalled;
function resetConfig() {
    const success = (0, node_child_process_1.spawnSync)('rm', ['-rf', exports.CLOUDFLARED_USR_DIR]).status === 0;
    if (!success)
        throw new Error('Failed to reset config, please delete .cloudflared folder manually.');
}
exports.resetConfig = resetConfig;
function getConfig() {
    return JSON.parse((0, node_fs_1.readFileSync)((0, node_path_1.resolve)(exports.CLOUDFLARED_USR_DIR, exports.TUNNELNAME), { encoding: 'utf-8' }).toString());
}
exports.getConfig = getConfig;
function isCredentialsFilesExist() {
    // check if .cloudflared folder exists
    if (!(0, node_fs_1.existsSync)(exports.CLOUDFLARED_USR_DIR))
        return false;
    const files = (0, node_fs_1.readdirSync)(exports.CLOUDFLARED_USR_DIR);
    // check if any of the files is missing
    return ['cert.pem', exports.TUNNELNAME].every((file) => files.includes(file));
}
exports.isCredentialsFilesExist = isCredentialsFilesExist;
function isTunnelCreated() {
    return getConfig().tunnelCreated;
}
exports.isTunnelCreated = isTunnelCreated;
function isAllReady() {
    return ['domain', 'tunnelCreated', 'tunnelId', 'routeDns'].every((key) => Boolean(getConfig()[key]));
}
exports.isAllReady = isAllReady;
function updateConfig(config) {
    if (!(0, node_fs_1.existsSync)(exports.CONFIG_FILE))
        (0, node_fs_1.writeFileSync)(exports.CONFIG_FILE, JSON.stringify({}), { encoding: 'utf-8' });
    (0, node_fs_1.writeFileSync)(exports.CONFIG_FILE, JSON.stringify({ ...getConfig(), ...config }), { encoding: 'utf-8' });
}
exports.updateConfig = updateConfig;
function isRemoteTunnelExist() {
    const result = (0, node_child_process_1.spawnSync)(cloudflared_1.bin, ['tunnel', 'list']);
    let tunnel;
    if (result.status === 0) {
        const tunnels = result.stdout.toString().trim().split('\n').slice(2);
        for (const line of tunnels) {
            const [id, name] = line.trim().split(/\s+/);
            if (name === exports.TUNNELNAME) {
                tunnel = { id, name };
                break;
            }
        }
        return tunnel;
    }
    throw new Error('Failed to check for existing tunnels.');
}
exports.isRemoteTunnelExist = isRemoteTunnelExist;
function createTunnel() {
    const tunnelExist = isRemoteTunnelExist();
    if (tunnelExist) {
        updateConfig({ tunnelCreated: true, tunnelId: tunnelExist.id });
        return;
    }
    const result = (0, node_child_process_1.spawnSync)(cloudflared_1.bin, ['tunnel', 'create', exports.TUNNELNAME], { stdio: 'inherit' });
    if (result.status !== 0)
        throw new Error('Failed to create tunnel.');
    const files = (0, node_fs_1.readdirSync)(exports.CLOUDFLARED_USR_DIR);
    // find a file with .json extension
    const tunnelId = files.find((file) => file.endsWith('.json'));
    updateConfig({ tunnelCreated: true, tunnelId });
}
exports.createTunnel = createTunnel;
const authorizeDomain = () => {
    let domain = '';
    let confirm = '';
    while (confirm.toLowerCase() !== 'y') {
        domain = (0, readline_sync_1.question)('Enter your authorized domain (e.g. example.com): \n');
        domain = domain.replace(/^(https?:\/\/)?(www\.)?/i, '').replace(/\/$/, ''); // remove http(s)://, www. and trailing slash
        confirm = (0, readline_sync_1.question)(`You entered ${domain}, is this correct? (y/n) \n`);
    }
    // add domain to config
    updateConfig({ domain });
};
function login() {
    if ((0, node_child_process_1.spawnSync)(cloudflared_1.bin, ['login'], { stdio: 'inherit' }).status === 0)
        authorizeDomain();
    else
        throw new Error('Failed to login.');
}
exports.login = login;
function routeDnsToTunnel() {
    const { domain, routeDns } = getConfig();
    if (routeDns)
        return;
    const result = (0, node_child_process_1.spawnSync)(cloudflared_1.bin, ['tunnel', 'route', 'dns', '-f', exports.TUNNELNAME, `${exports.TUNNELNAME}.${domain}`]);
    if (result.status === 0)
        updateConfig({ routeDns: true });
}
exports.routeDnsToTunnel = routeDnsToTunnel;
async function startTunnel({ port, host }) {
    routeDnsToTunnel();
    core_1.ux.action.start('Starting tunnel...');
    const { url, connections, child, stop: _stop } = (0, cloudflared_1.tunnel)({ '--url': `${host}:${port}`, tunnel: exports.TUNNELNAME });
    console.log(`\n===================================\nExposing: ${host}:${port}\nSecure tunnel URL: ${await url}\nConnections Ready! ${JSON.stringify(await Promise.all(connections))}\n===================================\n`);
    core_1.ux.action.stop('Tunnel started. Press Ctrl+C to stop.');
    stop = _stop;
    await new Promise(resolve => {
        child.on('exit', resolve);
    });
}
exports.startTunnel = startTunnel;
function stopTunnel() {
    if (!stop)
        return;
    stop();
    stop = undefined;
}
exports.stopTunnel = stopTunnel;
async function installBinary() {
    core_1.ux.action.start('Installing cloudflared binary...');
    const installResult = (await (0, cloudflared_1.install)(cloudflared_1.bin)
        .then(() => ({
        text: 'Installed cloudflared binary.',
        type: 'success',
    }))
        .catch(error => ({
        text: 'Failed to install cloudflared binary.',
        type: 'error',
        error: error,
    })));
    core_1.ux.action.stop(installResult.text);
    if (installResult.error)
        throw installResult.error;
}
exports.installBinary = installBinary;
